### PR TITLE
Issues/3234 add parallelLabels for blueprint/pipeline

### LINKED ISSUE #3234: [Refactor][pipeline] pipeline need a mode with higher controllability than PIPELINE_MAX_PARALLEL
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-devlake/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

The pipeline now supports PIPELINE_MAX_PARALLEL to limit running parallel. But it's controllability is not enough. I'll use the following example to support my argument:

there are 2 blueprints:
1. blueprint 1: [ [jira], [customJiraPlugin], [starrocks] ]
2. blueprint 2: [ [github(repo1), gitextractor(repo1)], [github(repo2), gitextractor(repo2)], [starrocks] ]

Now, if I set PIPELINE_MAX_PARALLEL=1, sometimes the blueprint cannot run immediately. If I set PIPELINE_MAX_PARALLEL >= 2, the plugin startrocks may run parallel and it will cause some errors.

So I think we can add a label in plugins to solve this problem.

```go
func (plugin Refdiff) PapallelLabels() map[string]string {
        return string[]{
            "file" // it means all plugin with the label "file" cannot run parallel
        }
}
func (plugin Icla) PapallelLabels() map[string]string {
        return string[]{
            "$pluginName" // it means this plugin cannot run parallel
        }
}
func (plugin Jira) PapallelLabels() map[string]string {
        return string[]{
            "$connectionId" // it means the plugin have the pluginName and same connection id cannot run parallel
        }
}
```
The plugins with the same label are controlled by the same lock.

If we add this, we will avoid another case.

The blueprint is like the following example:
`[ [jira1, jira2], [starrocks] ]`

```go
func (plugin Jira) PapallelLabels() map[string]string {
        return string[]{
            "$connectionId" // it means the plugin have the pluginName and same connection id cannot run parallel
        }
}
```
We can add magic labels like the above one to prevent jira1 and jira2 from using the same rate limit at the same time. 

The env `PIPELINE_MAX_PARALLEL` will use keep and it means how many pipelines can run parallel.
A new env `TASK_MAX_PARALLEL` will be added for how many tasks can run parallel.

Example:
```env
PIPELINE_MAX_PARALLEL=1
TASK_MAX_PARALLEL=2
```
It means 1 pipeline can run together and at most 2 tasks at the same stage in the running pipeline can be started.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by likyh**
### Summary
1. add labels for blueprint/pipeline
2. use labels which start with `parallel/*` to control how pipeline run parallel.
3. [config-ui] use api `trigger` to run blueprint
4. delete unused param `retry`
5. fix a bug imported by https://github.com/apache/incubator-devlake/pull/3754: the response body from `/blueprints` changed.

image

Now the pipelines with the same labels which start with `parallel/*` (just need one) can only run one by one.
all labels in the blueprint will pass to the pipeline when triggered.

### Does this close any open issues?
Closes #3234

### Screenshots
image
image
image

image
parallel_label in `Get /pipelines` must be a complete label name to filter.
image

### CODE REVIEW

File: models/migrationscripts/20221115_add_parallel_labels.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+type addParallelLabels struct{}
+
+func (*addParallelLabels) Up(res core.BasicRes) errors.Error {
+	db := res.GetDal()
+	err := db.AutoMigrate(&DbPipelineParallelLabel20221115{})
+	if err != nil {
+		return err
+	}
+	err = db.AutoMigrate(&DbBlueprintParallelLabel20221115{})
```
Comment (warren830) [inline models/migrationscripts/20221115_add_parallel_labels.go:56]: why there are two automigrate for the same table?

File: models/migrationscripts/20221115_add_parallel_labels.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	}
+	return nil
+}
+
+func (*addParallelLabels) Version() uint64 {
+	return 20221115000034
+}
+
+func (*addParallelLabels) Name() string {
+	return "UpdateSchemas for addParallelLabels"
```
Comment (warren830) [inline models/migrationscripts/20221115_add_parallel_labels.go:68]: `add _devlake_blueprint_parallel_labels` is better?

Comment (warren830) [inline models/migrationscripts/20221115_add_parallel_labels.go:68]: what about `add parallel labels' schema for blueprint and pipeline`

Comment (likyh) [reply models/migrationscripts/20221115_add_parallel_labels.go]: fixed

**Review by mappjzc — DISMISSED** [review]

File: models/migrationscripts/20221115_add_parallel_labels.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+type addLabels struct{}
+
+func (*addLabels) Up(res core.BasicRes) errors.Error {
+	db := res.GetDal()
+	err := db.AutoMigrate(&DbPipelineLabel20221115{})
+	if err != nil {
+		return err
+	}
+	err = db.AutoMigrate(&DbBlueprintLabel20221115{})
```
[the remark this answers is not in the document]
Comment (likyh) [reply models/migrationscripts/20221115_add_parallel_labels.go:56]: good idea

File: services/pipeline.go [hunk 4]
```diff
@@ -165,29 +167,57 @@ func RunPipelineInQueue(pipelineMaxParallel int64) {
...
+				pipelineParallelLabels = append(pipelineParallelLabels, dbLabel.Name)
+			}
+		}
+		runningParallelLabels = append(runningParallelLabels, pipelineParallelLabels...)
+
+		go func(pipelineId uint64, parallelLabels []string) {
 			defer sema.Release(1)
-			globalPipelineLog.Info("run pipeline, %d", pipelineId)
+			defer func() {
+				runningParallelLabels = utils.SliceRemove(runningParallelLabels, parallelLabels...)
```
[the remark this answers is not in the document]
Comment (likyh) [reply services/pipeline.go:217]: fix, add lock for this

**Review by mintsweet — APPROVED** [review]
LGTM.

**Review by klesh — DISMISSED** [review]
LGTM

File: services/pipeline.go [hunk 5]
```diff
@@ -167,27 +171,58 @@ func RunPipelineInQueue(pipelineMaxParallel int64) {
 		dbPipeline := &models.DbPipeline{}
 		for {
 			cronLocker.Lock()
+			// prepare query to find an appropriate pipeline to execute
 			db.Where("status IN ?", []string{models.TASK_CREATED, models.TASK_RERUN}).
+				Joins(`left join _devlake_pipeline_labels ON
+                  _devlake_pipeline_labels.pipeline_id = _devlake_pipelines.id AND
```
[the remark this answers is not in the document]
Comment (likyh) [reply services/pipeline.go]: fixed

**Review by klesh — APPROVED** [review]
LGTM

**Pull request merged**